MigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated[backups:3,nodes:4,fragmented:true]

_master_ (commit c6299681f3f6803124dcf8f10e4dcf1d2460cfb1)

Failed on ZuluJDK11 : http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-ZuluJDK11/46/testReport/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_backups_3_nodes_4_fragmented_true_/

**Stacktrace**

```
java.lang.AssertionError: 
Partition: 82 is leaking on [127.0.0.1]:5712 - 21a918c1-1ac5-403e-aec8-1b293b255344
Expected: (a value equal to or greater than <0> and a value less than or equal to <3>)
     but: a value less than or equal to <3> <6> was greater than <3>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertNoLeakingData(PartitionCorrectnessTestSupport.java:270)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:238)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.access$000(PartitionCorrectnessTestSupport.java:63)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:389)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1308)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1410)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:206)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:202)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:98)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:77)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)

```

```
Hiccups measured while running test 'testPartitionData_whenBackupNodesStartedTerminated[backups:3,nodes:4,fragmented:true](com.hazelcast.internal.partition.MigrationCorrectnessTest):'
08:58:55, accumulated pauses: 277 ms, max pause: 137 ms, pauses over 1000 ms: 0
08:59:00, accumulated pauses: 98 ms, max pause: 38 ms, pauses over 1000 ms: 0
08:59:05, accumulated pauses: 53 ms, max pause: 27 ms, pauses over 1000 ms: 0
08:59:10, accumulated pauses: 25 ms, max pause: 0 ms, pauses over 1000 ms: 0
08:59:15, accumulated pauses: 25 ms, max pause: 1 ms, pauses over 1000 ms: 0
08:59:20, accumulated pauses: 34 ms, max pause: 12 ms, pauses over 1000 ms: 0
08:59:25, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
08:59:30, accumulated pauses: 25 ms, max pause: 0 ms, pauses over 1000 ms: 0
08:59:35, accumulated pauses: 50 ms, max pause: 16 ms, pauses over 1000 ms: 0
08:59:40, accumulated pauses: 32 ms, max pause: 1 ms, pauses over 1000 ms: 0
08:59:45, accumulated pauses: 40 ms, max pause: 3 ms, pauses over 1000 ms: 0
08:59:50, accumulated pauses: 57 ms, max pause: 25 ms, pauses over 1000 ms: 0
08:59:55, accumulated pauses: 30 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:00:00, accumulated pauses: 53 ms, max pause: 3 ms, pauses over 1000 ms: 0
09:00:05, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:00:10, accumulated pauses: 38 ms, max pause: 10 ms, pauses over 1000 ms: 0
09:00:15, accumulated pauses: 29 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:00:20, accumulated pauses: 27 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:00:25, accumulated pauses: 51 ms, max pause: 16 ms, pauses over 1000 ms: 0
09:00:30, accumulated pauses: 30 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:00:35, accumulated pauses: 26 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:00:40, accumulated pauses: 37 ms, max pause: 12 ms, pauses over 1000 ms: 0
09:00:45, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:00:50, accumulated pauses: 30 ms, max pause: 1 ms, pauses over 1000 ms: 0
09:00:55, accumulated pauses: 60 ms, max pause: 24 ms, pauses over 1000 ms: 0
09:01:00, accumulated pauses: 39 ms, max pause: 2 ms, pauses over 1000 ms: 0
09:01:05, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
09:01:10, accumulated pauses: 46 ms, max pause: 13 ms, pauses over 1000 ms: 0
09:01:15, accumulated pauses: 341 ms, max pause: 335 ms, pauses over 1000 ms: 0
```

Failure on maintenance : 

http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.maintenance-ZingJDK7/91/testReport/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withRestart_backups_3_nodes_4_fragmented_false_/


**Stacktrace**

```
java.lang.AssertionError: 
Partition: 60 is leaking on [127.0.0.1]:5701 - 8c143315-28ed-4d8a-96ee-ae0e172ade27
Expected: (a value equal to or greater than <0> and a value less than or equal to <3>)
     but: a value less than or equal to <3> <5> was greater than <3>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertNoLeakingData(PartitionCorrectnessTestSupport.java:270)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:238)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.access$000(PartitionCorrectnessTestSupport.java:63)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:389)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1310)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1412)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:206)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:202)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withRestart(AbstractMigrationCorrectnessTest.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.internal.partition.impl;
 
+import com.hazelcast.cluster.Address;
 import com.hazelcast.cluster.ClusterState;
 import com.hazelcast.cluster.Member;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
@@ -24,8 +25,10 @@ import com.hazelcast.core.OperationTimeoutException;
 import com.hazelcast.instance.impl.Node;
 import com.hazelcast.internal.cluster.impl.ClusterServiceImpl;
 import com.hazelcast.internal.metrics.Probe;
+import com.hazelcast.internal.partition.IPartitionLostEvent;
 import com.hazelcast.internal.partition.InternalPartition;
 import com.hazelcast.internal.partition.InternalPartitionService;
+import com.hazelcast.internal.partition.MigrationEndpoint;
 import com.hazelcast.internal.partition.MigrationInfo;
 import com.hazelcast.internal.partition.MigrationInfo.MigrationStatus;
 import com.hazelcast.internal.partition.MigrationStateImpl;
@@ -48,7 +51,6 @@ import com.hazelcast.internal.util.collection.Int2ObjectHashMap;
 import com.hazelcast.internal.util.collection.PartitionIdSet;
 import com.hazelcast.internal.util.scheduler.CoalescingDelayedTrigger;
 import com.hazelcast.logging.ILogger;
-import com.hazelcast.cluster.Address;
 import com.hazelcast.partition.ReplicaMigrationEvent;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 import com.hazelcast.spi.impl.InternalCompletableFuture;
@@ -57,14 +59,13 @@ import com.hazelcast.spi.impl.executionservice.ExecutionService;
 import com.hazelcast.spi.impl.operationservice.Operation;
 import com.hazelcast.spi.impl.operationservice.OperationService;
 import com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl;
-import com.hazelcast.internal.partition.IPartitionLostEvent;
-import com.hazelcast.internal.partition.MigrationEndpoint;
 import com.hazelcast.spi.properties.ClusterProperty;
 import com.hazelcast.spi.properties.HazelcastProperties;
 
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.LinkedHashSet;
@@ -75,6 +76,7 @@ import java.util.Objects;
 import java.util.Queue;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
@@ -121,6 +123,7 @@ public class MigrationManager {
     private final boolean fragmentedMigrationEnabled;
     private final long memberHeartbeatTimeoutMillis;
     private boolean triggerRepartitioningWhenClusterStateAllowsMigration;
+    private final Set<MigrationInfo> finalizingMigrationsRegistry = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
     MigrationManager(Node node, InternalPartitionServiceImpl service, Lock partitionServiceLock) {
         this.node = node;
@@ -213,8 +216,8 @@ public class MigrationManager {
 
                 MigrationEndpoint endpoint = source ? MigrationEndpoint.SOURCE : MigrationEndpoint.DESTINATION;
                 FinalizeMigrationOperation op = new FinalizeMigrationOperation(migrationInfo, endpoint, success);
-                op.setPartitionId(partitionId).setNodeEngine(nodeEngine).setValidateTarget(false)
-                        .setService(partitionService);
+                op.setPartitionId(partitionId).setNodeEngine(nodeEngine).setValidateTarget(false).setService(partitionService);
+                registerFinalizingMigration(migrationInfo);
                 OperationServiceImpl operationService = nodeEngine.getOperationService();
                 if (operationService.isRunAllowed(op)) {
                     // When migration finalization is triggered by subsequent migrations
@@ -242,6 +245,18 @@ public class MigrationManager {
         } finally {
             migrationInfo.doneProcessing();
         }
+    }
+
+    private void registerFinalizingMigration(MigrationInfo migration) {
+        finalizingMigrationsRegistry.add(migration);
+    }
+
+    public boolean removeFinalizingMigration(MigrationInfo migration) {
+        return finalizingMigrationsRegistry.remove(migration);
+    }
+
+    public boolean isFinalizingMigrationRegistered(int partitionId) {
+        return finalizingMigrationsRegistry.stream().anyMatch(m -> partitionId == m.getPartitionId());
     }
 
     /**

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
@@ -112,6 +112,15 @@ abstract class BaseMigrationOperation extends AbstractPartitionOperation
         if (!partitionService.applyCompletedMigrations(completedMigrations, migrationInfo.getMaster())) {
             throw new PartitionStateVersionMismatchException(partitionStateVersion, partitionService.getPartitionStateVersion());
         }
+        if (partitionService.getMigrationManager().isFinalizingMigrationRegistered(migrationInfo.getPartitionId())) {
+            // There's a pending migration finalization operation in the queue.
+            // This happens when this node was the source of a backup replica migration
+            // and now it is destination of another replica migration on the same partition.
+            // Sources of backup migrations are not part of migration transaction
+            // and they learn the migration only while applying completed migrations.
+            throw new RetryableHazelcastException("There is a scheduled FinalizeMigrationOperation for the same partition => "
+                    + migrationInfo);
+        }
     }
 
     /** Verifies that the sent partition state version matches the local version or this node is master. */

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizeMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizeMigrationOperation.java
@@ -72,6 +72,10 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
         PartitionStateManager partitionStateManager = partitionService.getPartitionStateManager();
         int partitionId = migrationInfo.getPartitionId();
 
+        if (!partitionService.getMigrationManager().removeFinalizingMigration(migrationInfo)) {
+            throw new IllegalStateException("This migration is not registered as finalizing: " + migrationInfo);
+        }
+
         if (isOldBackupReplicaOwner() && partitionStateManager.isMigrating(partitionId)) {
             // On old backup replica, migrating flag is not set during migration.
             // Because replica is copied from partition owner to new backup replica.
@@ -79,11 +83,8 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
             // and completed migration is published by master.
             // If this partition's migrating flag is set, then it means another migration
             // is submitted to this member for the same partition and it's already executed.
-            // This finalization is now obsolete.
-            getLogger().fine("Cannot execute migration finalization, because this member was previous owner of a backup replica,"
-                    + " and a new migration operation has already superseded this finalization. "
-                    + "This operation is now obsolete. -> " + migrationInfo);
-            return;
+            throw new IllegalStateException("Another replica migration is started on the same partition before finalizing "
+                    + migrationInfo);
         }
 
         NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
@@ -185,8 +186,7 @@ public final class FinalizeMigrationOperation extends AbstractPartitionOperation
                         + partitionId);
             }
         } else {
-            int replicaOffset = migrationInfo.getDestinationCurrentReplicaIndex() <= 1 ? 1 : migrationInfo
-                    .getDestinationCurrentReplicaIndex();
+            int replicaOffset = Math.max(migrationInfo.getDestinationCurrentReplicaIndex(), 1);
 
             for (ServiceNamespace namespace : replicaManager.getNamespaces(partitionId)) {
                 long[] versions = updatePartitionReplicaVersions(replicaManager, partitionId, namespace, replicaOffset - 1);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationOperation.java
@@ -142,7 +142,7 @@ public class MigrationOperation extends BaseMigrationOperation implements Target
             InternalPartitionServiceImpl partitionService = getService();
             PartitionReplicaManager replicaManager = partitionService.getReplicaManager();
             int destinationNewReplicaIndex = migrationInfo.getDestinationNewReplicaIndex();
-            int replicaOffset = destinationNewReplicaIndex <= 1 ? 1 : destinationNewReplicaIndex;
+            int replicaOffset = Math.max(destinationNewReplicaIndex, 1);
 
             Map<ServiceNamespace, long[]> namespaceVersions = fragmentMigrationState.getNamespaceVersionMap();
             for (Entry<ServiceNamespace, long[]> e  : namespaceVersions.entrySet()) {

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/AbstractGracefulShutdownCorrectnessTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/AbstractGracefulShutdownCorrectnessTest.java
@@ -38,6 +38,7 @@ import java.util.concurrent.TimeUnit;
 
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 
@@ -147,6 +148,7 @@ public abstract class AbstractGracefulShutdownCorrectnessTest extends PartitionC
 
         for (int p = 0; p < partitionCount; p++) {
             Integer actual = (Integer) operationService.invokeOnPartition(null, new TestGetOperation(), p).join();
+            assertNotNull(actual);
             assertEquals(value, actual.intValue());
         }
     }

